// build/src/task/run.js
'use strict';

var argv = require('yargs').argv;
var run = require('run');

var delimiter = process.platform === 'win32' ? ';' : ':';

module.exports = require('../task').extend({
  name: 'run',
  execute: function(done) {
    process.env['NODE_PATH'] = process.env['NODE_PATH'] || '';
    process.env['NODE_PATH'] += delimiter +
      ['.', './bin/src', './bin/' + this.parent.name + '/src'].join(delimiter);

    var args = ['bin/' + this.parent.name + '/main.js'];

    if(argv.config && argv.config !== true) {
      args.push(argv.config);
    }

    var command = {
      name: 'node',
      args: args
    };

    if(argv.forever) {
      command.args = [];

      if(typeof argv.forever === 'string') {
        command.args = argv.forever.split(/\s+/);
      }

      command.args.push(command.name);
      command.args = command.args.concat(args);
      command.name = 'forever';
    }

    run(command, null, {
      cwd: process.cwd(),
      env: process.env,
      next: done
    });
  }
}, { export: 'Run' });
