// build/src/util/path.js
'use strict';

var _dir = __dirname;
var _path = require('path');

var Path = function(path, root) {
  this.fragments = path ? Array.isArray(path) ? path : [path] : [];
  this.root = root || _dir + '/../..';
};

Path.prototype = {
  append: function(fragment) {
    this.fragments.push(fragment);
    return this;
  },
  pop: function() {
    this.fragments.pop();
    return this;
  },
  abs: function() {
    var fragments = this.fragments.slice(0);
    fragments.unshift(this.root);
    return Path.join.apply(Path,
                           fragments.concat(Array.prototype.slice.call(arguments, 0)));
  },
  rel: function() {
    return './' + Path.join.apply(Path,
                           this.fragments.concat(Array.prototype.slice.call(arguments, 0)));
  }
};

Path.join = function() {
  var args = arguments && arguments.length &&
      Array.prototype.slice.call(arguments, 0) || [];
  var all = [];

  for(var i = 0, length = args.length; i < length; ++i) {
    var arg = args[i];
    all = Array.isArray(arg) ? all.concat(arg) : all.concat(arg.split(_path.sep));
  }

  return _path.normalize(all.join(_path.sep));
};

Path.extend = require('class-extend').extend;

module.exports = Path;
